### ISSUE TITLE
Inconsistency between overlapping frames

### ISSUE THREAD

**Issue by sssangha** [insarlab/PySolid#54]
When using `calc_solid_earth_tides_grid` through [MintPy](https://github.com/insarlab/MintPy/blob/7704c34797a5db62d2521e84010561ac8cbf94f2/src/mintpy/solid_earth_tides.py#L164-L170) for two separate runs involving overlapping frames, I notice unexpected behavior.

Specifically, I would expect the overlapping region between the two to be consistent after referencing epochs in the stack to the same reference date and point, but I instead find a non-zero, sub-mm level mean offset (~0.1 mm, and a range of ~0.5mm) and stdev (~0.33 mm):
![image](https://user-images.githubusercontent.com/13227405/226149356-f0508999-bb15-4ed5-b925-f28bc7a76555.png)

Here is the residual within the overlapping region:
![image](https://user-images.githubusercontent.com/13227405/226149369-c07308fd-5355-46f6-bd0c-65af6b8df654.png)

Please let me know how I can help to debug, if I should clarify anything, and if also perhaps it would be best to share the input files I used on the MintPy end. Many thanks

**Comment by yunjunz** [comment]
Hmm, I would expect the overlapping region having the same SET as well. The 0.5 mm is really small, I am wondering if the two frames have the geometry file from the same date as well?

And similarly, could you try to concatenate the incidence and azimuth angle, and check their overlapping regions too?

**Comment by sssangha** [comment]
Hello @yunjunz upon closer inspection, it looks like the geometry files indeed are slightly off between the two frames, so I wonder if this may 
![Screenshot 2023-03-22 at 7 12 36 PM](https://user-images.githubusercontent.com/13227405/227318919-2bce133a-0085-4efd-a9d1-4b017d43a50d.png)
![Screenshot 2023-03-22 at 7 13 50 PM](https://user-images.githubusercontent.com/13227405/227318924-44a68068-1c40-4807-a4c9-f31f2eaa4df3.png)
be associated with resampling of the geometry files themselves:


Regardless of this though, as a rough sanity check I hacked the `solid_earth_tide.py` script through mintpy to directly inspect the outputs of `calc_solid_earth_tides_grid` (i.e. the direct E, N, and U outputs). So i.e. I took this [section of the code](https://github.com/insarlab/MintPy/blob/7704c34797a5db62d2521e84010561ac8cbf94f2/src/mintpy/solid_earth_tides.py#L174-L176) and overrode it to iteratively pass the E, N, and U components into separate cubes. What I found after doing so was still a shift between the frames (see the E component below). I understand this is not to be expected, and am struggling to pinpoint the source of the discrepancy:
![Screenshot 2023-03-23 at 11 46 28 AM](https://user-images.githubusercontent.com/13227405/227318892-4f8197c9-e03c-4b69-892f-47bdff81fa10.png)

**Comment by yunjunz** [comment]
I guess this could be due to the `step_size` used in the SET calculation for speedup. Could you test the `set_e` again by setting `step_size = 1` [here](https://github.com/insarlab/MintPy/blob/7704c34797a5db62d2521e84010561ac8cbf94f2/src/mintpy/solid_earth_tides.py#L156)?

If confirmed, it means we should adjust the [step/grid selection](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L55-L74) and [resize](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L106-L113) code .

**Comment by sssangha** [comment]
@yunjunz , your suggested test works!
![Screenshot 2023-03-27 at 1 50 23 PM](https://user-images.githubusercontent.com/13227405/228063558-2071dd01-1f88-465d-8e4b-e3c6c54cacaa.png)

So looks like we would have to adjust the step as you advised. Thanks again for helping to debug!

**Comment by sssangha** [comment]
Hello @yunjunz , as an amendment to my previous comment I manually re-referenced the two frames to the same coordinate after the fields were generated. To clarify though, is this necessary or should the outputs be agnostic to a reference point? I see the output `SET.h5` cubes themselves had different recorded reference points, but it isn't clear whether or not they were applied to the output array itself.

**Comment by sssangha** [comment]
If subtracting 1 from the dimensions [like this](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L87-L88) intentional?

**Comment by yunjunz** [comment]
> If subtracting 1 from the dimensions [like this](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L87-L88) intentional?

Yes, this is intentional, to ensure the generated SET from the fortran program to have the size of `length x width`.

**Comment by yunjunz** [comment]
> To clarify though, is this necessary or should the outputs be agnostic to a reference point? I see the output SET.h5 cubes themselves had different recorded reference points, but it isn't clear whether or not they were applied to the output array itself.

Assuming this is still `set_e` (not the SET projected into the LOS direction), it should not need a reference point, as everything is absolute. In mintpy, the referencing in time/space while correcting for SET is done on the fly.

> Because if I don't re-reference the grids manually, I get a constant offset as so, event after setting the step_size to 1:

Hmm... Is the magnitude of this offset equivalent to shifting one or a few pixels?

**Comment by yunjunz** [comment]
[Thoughts In Progress] until we figure out the offset without spatial referencing.

With these tests you have done here, which is super helpful, I think we should adjust the step as below, e.g. given a shape of 1033 x 2526 with a step size of 20:

1. coarse grid starting coord should be (10, 10), instead of (0, 0).
2. number of steps should be `ceil(1033 / 20) = 52` and `ceil(2526 / 20) = 127`, instead of using `rint()`
3. resize the coarse resolution to the original resolution with a shape of (52 x 20, 127 x 20)
4. crop the original resolution to the original shape of (1033, 2526)

**Comment by sssangha** [comment]
> Hmm... Is the magnitude of this offset equivalent to shifting one or a few pixels?

Apologies, not sure I understand. It's a vertical offset that doesn't seem to translate laterally. And yes I did manually hack it still to use `set_e`

Thanks for the advice here, I'm working to test out your suggestions. Quick point of clarification though, how would you be defining the starting coord using the step_size in this case? Would it be like:
```
lat0 = float(atr['Y_FIRST'])
lat0 += float(atr['Y_STEP']) * step_size
lon0 = float(atr['X_FIRST'])
lon0 +=  float(atr['X_STEP']) * step_size
```


> number of steps should be ceil(1033 / 20) = 52 and ceil(2526 / 20) = 127, instead of using rint()
So I would adjust this e.g. as so?:
```
    length = np.ceil(int(atr['LENGTH']) / num_step).astype(int)
    width  = np.ceil(int(atr['WIDTH'])  / num_step).astype(int)
```

**Comment by sssangha** [comment]
@yunjunz , I just realized something. The fact that we don't use the same datetime info between these frames should play a role here, no? E.g. refer to [this section ](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L85-L86)where we feed in this info to the function.

**Comment by sssangha** [comment]
Ok, so it looks like I may be onto something with my hunch above. So for context we are working to integrate these SET estimates into our ARIA GUNW products as we had discussed a few months ago. With that in mind, after fixing the times between adjacent frames, I'm finally above to resolve this apparent discrepancy between frames:
![Screenshot 2023-03-28 at 8 58 46 PM](https://user-images.githubusercontent.com/13227405/228423550-cba080c9-522a-4010-8f16-71c1dcfc6cd0.png)


So in other words, it looks like the specified timing somehow introduces a vertical offset. It's not obvious to me why this is so.

**Comment by yunjunz** [comment]
> And also, how would this work if the specified step_size = 1? You wouldn't have a coarse grid, but would still shift the starting lat/lon by a pixel, unless you introduce an if/else statement to prevent this?

Good point. Then we should still use the precise corner coordinates of the corner pixels, i.e. the UL pixel's UL corner, UR pixel's UR corner, LL pixel's LL corner and LR pixel's LR corner. The adjusted procedure should be, e.g. given a shape of 1033 x 2526 with a step size of 20:

1. coarse grid starting coordinates: `Y_FIRST` and `X_FIRST` 
2. number of steps should be `ceil(1033 / 20) + 1 = 53` and `ceil(2526 / 20) + 1 = 128`
3. resize the coarse resolution to the original resolution with a shape of (52 x 20, 127 x 20), not (53 x 20, 128 x 20).
4. crop the original resolution to the original shape of (1033, 2526)

#

> crop the original resolution to the original shape of (1033, 2526). This would already be captured by this [section](https://github.com/insarlab/PySolid/blob/302ac2c89590b15b7ba81c368b8e8c284cdbaf10/src/pysolid/grid.py#L111-L113), right?

That does not capture the cropping, at least not properly accounting for the extra pixels falling into the last partial coarse step.

**Comment by sssangha** [comment]
> I intended to write a `swath` mode, in addition to the current `grid` and `point` mode, for this scenario. The implementation will be much easier if we have a true API-like talking between the python and fortran code (#2), instead of the current talking via text file. Therefore, I have not done it yet.

To clarify though, how do you envision this `swath` mode? Would you increment through multiple frames individually? From what I understand, merged products passed either through ISCE or ARIA only report one time corresponding to the center of the scene, so how would you account for this change in timing along track other than making some assumptions or perhaps recording/passing sensing start/stop times?

> Note that adding a constant offset, or setting the same UTC time, fixes the "look" (to be continuous), but the ramp along azimuth still exists, which we have been ignoring so far.

So for the aforementioned application involving computing and embedding SET as downsampled layers for 4 height levels within the GUNW, could you please advise on my current plan of action, specifically on whether or not this sounds good on a high level? We compute SET layers for each [individual frame](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/pull/124), for which of course different sensing times for the center of the scenes are captured and passed. This leads to this apparent offset which I "correct" by sequentially stitching in cases multiple frames are involved (i.e. estimate the integer offset in areas of overlap at incrementally apply it). This is captured in this [ARIA-tools PR](https://github.com/aria-tools/ARIA-tools/pull/353).


As far as commits to pysolid, I will issue a PR capturing your most recent suggestions regarding adjustments to the number of steps. Perhaps from there we could perhaps gameplan how a `swath` mode would work, depending of course on our bandwidth! What do you think?

Again thanks for your ongoing help and advice, myself and the team really appreciate it!

**Comment by sssangha** [comment]
> Your plan of action looks good to me, except for one detail: the output of pysolid is already interpolated, so I would leverage that, and stitch the interpolated SET in ARIA-tools, instead of the downsampled one.

Great, but to clarify I meant to stitch the downsampled products intentionally for the sake of efficiency/speed. I.e. current interpolation procedure takes ~a few secs or so depending on the footprint of the image, so it would save a great deal of time to interpolate after stitching. Unless you would recommend otherwise?

> The start/stop sensing time information is available from ISCE (I am not sure about this detail in ARIA products), which can be used to generate the pixel-wised 2D matrix for the timing, similar to what we have for incidence angle. We could have a function in pysolid or mintpy to generate this.

Are you referring to [this function](https://github.com/insarlab/MintPy/blob/7704c34797a5db62d2521e84010561ac8cbf94f2/src/mintpy/utils/utils0.py#L98)?

**Comment by yunjunz** [comment]
> Great, but to clarify I meant to stitch the downsampled products intentionally for the sake of efficiency/speed. I.e. current interpolation procedure takes ~a few secs or so depending on the footprint of the image, so it would save a great deal of time to interpolate after stitching. Unless you would recommend otherwise?

I don't have a strong opinion on this: code simplicity and efficiency are both important. Your call.

> Are you referring to [this function](https://github.com/insarlab/MintPy/blob/7704c34797a5db62d2521e84010561ac8cbf94f2/src/mintpy/utils/utils0.py#L98)?

Yes.
